Add support for reading cgroup.* files

drakenclimber opened the pull request:
Add support to `cgroup_get_cgroup()` to read disabled controllers and psuedo-controllers like cgroup.*.  Note that this change was only added to cgroup v2 and cgroup v1 remains the same.

Example usage:
```
#!/usr/bin/env python3
from libcgroup import Cgroup, Version

# Read all _enabled_ controllers.  The "cgroup" controller will not be read in this case
cg = Cgroup('user.slice', Version.CGROUP_V2)
cg.get()
print(cg)

# Read the cpuset controller.  This will work whether it's enabled or disabled
cg = Cgroup('user.slice', Version.CGROUP_V2)
cg.add_controller('cpuset')
cg.get()
print(cg)

# Read the "cgroup" controller
cg = Cgroup('user.slice', Version.CGROUP_V2)
cg.add_controller('cgroup')
cg.get()
print(cg)

# And of course, it's legal to combine controllers as you see fit
cg = Cgroup('user.slice', Version.CGROUP_V2)
cg.add_controller('cpuset')
cg.add_controller('cgroup')
cg.add_controller('memory')
cg.get()
print(cg)
```

Closes: #297

drakenclimber commented on src/api.c:
```diff
@@ -5267,6 +5283,18 @@ int cgroup_get_controller_next(void **handle, struct cgroup_mount_point *info)
 		goto out_unlock;
 	}
 
+	if (strncmp(cg_mount_table[*pos].name, CGROUP_FILE_PREFIX, CONTROL_NAMELEN_MAX) == 0)
```
This change was needed to appease `lscgroup`.  The `cgroup` controller was showing up in its lists of controllers, and I don't think this is desired behavior

drakenclimber commented:
@justinfx - Do you want to weigh in on these changes before they're merged?

justinfx commented:
@drakenclimber thanks for working on this!
It looks great to me from a code standpoint and I understand and agree with hiding the pseudo cgroup controller from listing.

I was wondering if this change might fix another problem I was seeing as well. I was finding that libcgroup seemed to report a group as not existing if I try to load one that has no subtree_control entries. As a specific example, if I created a parent group and configured controllers and then added a child "leaf" with no controllers so that I can add pids without the "no inner process" failure, libcgroup seemed to say "leaf" doesn't exist. I then had to at least add the "cpu" controller to the "leaf" subtree_control to make the group loadable. Does that sound like a problem that might exist and might be fixed by the changes here where it will load all controllers anyways?

drakenclimber commented:
> @drakenclimber thanks for working on this! It looks great to me from a code standpoint and I understand and agree with hiding the pseudo cgroup controller from listing.

Thanks!  Glad to help :)

> I was wondering if this change might fix another problem I was seeing as well. I was finding that libcgroup seemed to report a group as not existing if I try to load one that has no subtree_control entries. As a specific example, if I created a parent group and configured controllers and then added a child "leaf" with no controllers so that I can add pids without the "no inner process" failure, libcgroup seemed to say "leaf" doesn't exist. I then had to at least add the "cpu" controller to the "leaf" subtree_control to make the group loadable. Does that sound like a problem that might exist and might be fixed by the changes here where it will load all controllers anyways?

Do you have an example program that exhibits the problem?  I just ran the following on the HEAD of libcgroup (without the changes in this patchset), and it worked for me, but I'm guessing I am missing something.

```
#!/usr/bin/env python3
from libcgroup import Cgroup, Version
import time

parentcg = Cgroup('parent', Version.CGROUP_V2)
parentcg.create()

childcg = Cgroup('parent/child', Version.CGROUP_V2)
childcg.create()

childcg.attach()

time.sleep(100)
```

The sleep was simply there to give me time to explore around in bash and verify that the python process was present in `parent/child/cgroup.procs`

drakenclimber commented:
Oh, and do you want credit for reporting this issue, @justinfx?  If so, please provide the following:

`Reported-by: FirstName LastName <youremailaddress@somewhere.com>`

Thanks!

justinfx commented:
> Oh, and do you want credit for reporting this issue, @justinfx? If so, please provide the following:

Sure!
`Reported-by: Justin Israel <justinisrael@gmail.com>`

> Do you have an example program that exhibits the problem?

Actually, yes! Your sample script can reproduce it with a small tweak to make it match my actual use case:

```
#!/usr/bin/env python3
from libcgroup import Cgroup, Version

parentcg = Cgroup('parent', Version.CGROUP_V2)
parentcg.create()

childcg = Cgroup('parent/child', Version.CGROUP_V2)
childcg.create()

childcg2 = Cgroup('parent/child', Version.CGROUP_V2)
childcg2.get()
# RuntimeError: cgroup_get_cgroup failed: 50002
```

The problem occurs when you have an existing child cgroup that has no controllers enabled because the parent did not put anything into its own `cgroup.subtree_control`. But the following does work:

```
#!/usr/bin/env python3
from libcgroup import Cgroup, Version

parentcg = Cgroup('parent', Version.CGROUP_V2)
parentcg.add_controller('cpu')
parentcg.add_controller('memory')
parentcg.add_controller('cpuset')
parentcg.create()

childcg = Cgroup('parent/child', Version.CGROUP_V2)
# need to give child at least one controller allowed from parent,
# in its own subtree_control
childcg.add_controller('cpu')
childcg.create()

childcg2 = Cgroup('parent/child', Version.CGROUP_V2)
childcg2.get()
```

justinfx commented:
@drakenclimber I've tested this branch and it seems my other issue is actually not fixed by this patch. Although I am not clear if it should be, or if its entirely unrelated and should be reported in a separate ticket? The reason it felt related is because it felt like your changes were indeed trying to address the issue of not loading all controllers, and this second issue being related to not having any controllers enabled and resulting in the cgroup acting like it doesn't exist.

drakenclimber commented:
> @drakenclimber I've tested this branch and it seems my other issue is actually not fixed by this patch. Although I am not clear if it should be, or if its entirely unrelated and should be reported in a separate ticket? The reason it felt related is because it felt like your changes were indeed trying to address the issue of not loading all controllers, and this second issue being related to not having any controllers enabled and resulting in the cgroup acting like it doesn't exist.

Thanks for the feedback.  This is a tough balancing act we're trying to walk, and I'm not exactly sure what the right answer is.  We're colliding with a few different things - C expects one thing, Python another, and of course cgroup v1 and v2 behave very differently.

Recently, a similar question [1] regarding the C version of `cgroup_get_cgroup()` came in.  We changed how we were writing to the subtree_control file and we updated `cgroup_get_cgroup()` to report only the enabled controllers.

Hmmm... what would you expect libcgroup to return in a case like this:
```
sudo mkdir -p /sys/fs/cgroup/foo/bar
```
```
#!/usr/bin/env python3
from libcgroup import Cgroup, Version

cg = Cgroup('foo/bar', Version.CGROUP_V2)
cg.get()
print(cg)
```

I don't think RuntimeError is correct, but it seems reasonable that nothing is returned from `cg.get()`.  Thoughts?

[1] https://github.com/libcgroup/libcgroup/issues/272

drakenclimber commented:
Alright, I thought about it for a bit and added one small patch.  Let me know what you think, @justinfx.  It makes the above scenario work for me.

Here's the new patch:
https://github.com/drakenclimber/libcgroup/commit/604a74171ea53d706aca2a133a3c5ae94982116b

justinfx commented:
This patch works great! And I think it's the right thing to do. I can now get back an actual object for the group, and see that it has no enabled controllers. But I can then have access to stats, properties, proc lists, etc. It makes sense to reflect the group as existing, because pids are allowed to live in that child cgroup regardless of not having any controllers enabled. Really, those controllers in subtree_control are for subtree delegation. A leaf group may not have any enabled so that it can contain the pids for a parent group that does have controllers enabled.

drakenclimber commented:
> This patch works great! And I think it's the right thing to do. I can now get back an actual object for the group, and see that it has no enabled controllers. But I can then have access to stats, properties, proc lists, etc. It makes sense to reflect the group as existing, because pids are allowed to live in that child cgroup regardless of not having any controllers enabled. Really, those controllers in subtree_control are for subtree delegation. A leaf group may not have any enabled so that it can contain the pids for a parent group that does have controllers enabled.

Awesome.  And I completely agree with your analysis.

Thanks for testing it out!  I also added your `Reported-by:` to the latest version of the patchset

drakenclimber commented:
Pushed to the main and release-3.0 branches.  Thanks, @justinfx and @kamalesh-babulal.
